ext/pcre/php_pcre.c

```diff
@@ -2417,37 +2420,45 @@ PHP_FUNCTION(preg_replace_callback_array)
 	fci.size = sizeof(fci);
 	fci.object = NULL;
 
-	ZEND_HASH_FOREACH_STR_KEY_VAL(Z_ARRVAL_P(pattern), str_idx, replace) {
-		if (str_idx) {
-			ZVAL_STR_COPY(&regex, str_idx);
-		} else {
+	ZEND_HASH_FOREACH_STR_KEY_VAL(pattern, str_idx_regex, replace) {
+		if (!str_idx_regex) {
 			php_error_docref(NULL, E_WARNING, "Delimiter must not be alphanumeric or backslash");
 			zval_ptr_dtor(return_value);
 			RETURN_NULL();
 		}
 
 		if (!zend_is_callable_ex(replace, NULL, 0, NULL, &fcc, NULL)) {
 			zend_string *callback_name = zend_get_callable_name(replace);
-			php_error_docref(NULL, E_WARNING, "'%s' is not a valid callback", ZSTR_VAL(callback_name));
+			zend_type_error("'%s' is not a valid callback", ZSTR_VAL(callback_name));
 			zend_string_release_ex(callback_name, 0);
-			zval_ptr_dtor(&regex);
-			zval_ptr_dtor(return_value);
-			ZVAL_COPY(return_value, subject);
-			return;
+			RETURN_THROWS();
 		}
 
 		ZVAL_COPY_VALUE(&fci.function_name, replace);
 
-		replace_count += preg_replace_func_impl(&zv, &regex, &fci, &fcc, subject, limit, flags);
-		if (subject != return_value) {
-			subject = return_value;
-		} else {
-			zval_ptr_dtor(return_value);
-		}
+		replace_count += preg_replace_func_impl(&zv, str_idx_regex, /* regex_ht */ NULL, &fci, &fcc,
+			subject_str, subject_ht,
+			limit, flags);
 
-		zval_ptr_dtor(&regex);
+		switch (Z_TYPE(zv)) {
+			case IS_NULL:
+				ZEND_ASSERT(subject_str != NULL);
+				zval_ptr_dtor(return_value);
+				break;
+			case IS_STRING:
+				ZEND_ASSERT(subject_str != NULL);
+				zend_string_release(subject_str);
+				subject_str = zend_string_copy(Z_STR(zv));
+				break;
+			case IS_ARRAY:
+				ZEND_ASSERT(subject_ht != NULL);
+				zend_hash_copy(subject_ht, Z_ARR(zv), NULL);
+				break;
+			EMPTY_SWITCH_DEFAULT_CASE()
+		}
```

Comment:
This code looks very dubious, I'm not convinced it's correct.

Reply:
Yeaaaaaah, it's the only way which looks sensible to me as ``zv`` can only be a string, an array, or null, so that seemed like one way to go about it as I still don't know how to replicate the previous zval replacement behaviour

Reply:
Oh, I've just bumped into the same issue when I wanted to add a type for this function. :/ I also have no idea what to do with this.

Reply:
One solution is to use the ZPP Union check but then recreate a temp zval manually for ``zv``